keps/sig-node/5018-dra-adminaccess/README.md

```diff
@@ -0,0 +1,687 @@
+    // set to false, otherwise it is enabled.
+    //
+    // +optional
+    // +featureGate=DRAAdminAccess
+    AdminAccess *bool
+}
+```
+
+Admin access to devices is a privileged operation because it grants users access
+to devices that are in use by other users. Drivers might also remove other
+restrictions when preparing the device.
+
+In Kubernetes 1.31 (before this KEP was introduced), an example validating
+admission policy
+[was provided](https://github.com/kubernetes/kubernetes/blob/4aeaf1e99e82da8334c0d6dddd848a194cd44b4f/test/e2e/dra/test-driver/deploy/example/admin-access-policy.yaml#L1-L11)
+which restricted access to this option. It was formerly the responsibility of
+cluster admins to ensure that such a policy was installed if the cluster
+shouldn't have allowed unrestricted access.
+
+Starting in Kubernetes 1.33 (when this KEP was introduced), a validation has
+been added to the REST storage layer to only authorize `ResourceClaim` or
+`ResourceClaimTemplate` with `adminAccess: true` requests if their namespace has
+the `kubernetes.io/dra-admin-access` label to only allow it for users with
+additional privileges. More time is needed to figure out how that should work,
+therefore the field is placed behind the `DRAAdminAccess` feature gate.
+
+The `DRAAdminAccess` feature gate controls whether users can set the
+`adminAccess` field to true when requesting devices. That is checked in the
+apiserver. In addition, the scheduler will not allocate claims with admin access
+when the feature is turned off. This covers the scenario that the field might
+have been set in Kubernetes v1.31 where the field was available uncondiionally
+or while the feature was enabled.
+
+### REST Storage Changes
+
+In both pkg/registry/resource/resourceclaim/strategy.go and
+pkg/registry/resource/resourceclaimtemplate/strategy.go, update `Validate` and
+`ValidateUpdate` to call a new `authorizedForAdmin` function to check if the
+request is authorized to set the `adminAccess` field to devices based on the DRA
+admin namespace label.
+
+### Kube-controller-manager Changes
+
+// TODO: what part of claim.Status.Allocation should be updated? e.g.
+AdminAccess is part of `DeviceRequestAllocationResult` but need to set it for
+each device?
+
+In pkg/controller/resourceclaim/controller.go, process requests in `handleClaim`
+function to prevent creation of `ResourceClaim` when the `ResourceClaimTemplate`
+has the `adminAccess` field while the feature gate is turned off.
+
+### Kube-scheduler Changes
+
+In pkg/scheduler/framework/plugins/dynamicresources/allocateddevices.go, handle
+claims with `adminAccess` to ensure devices allocated with `adminAccess` are
+skipped without invoking the callback bypassing standard allocation checks.
+
+### ResourceQuota
+
+Requests asking for `adminAccess` contribute to the quota. We recommend that
+these administrator namespaces should not have quotas configured.
+
```

Comment:
Two areas I am curious on.

Area 1: 
  Resource Quota and Kueue are very similar in this regard. I guess for Kueue one could ignore the count of devices coming from adminAccess resource claims.
  Is there any upper limit of the number of requests to a GPU than people could do? I wonder if it would be possible to have too many admin requests to a GPU and that would block the user's workload?
  
  Area 2:
  
  Is there any impact to autoscaling for adminAccess?

Reply:
> Area 1

Other components like Kueue should indeed follow the same resource tracking logic and ignore admin access when considering resource needs.

Because this is a privileged feature, the assumption is that admins use this judiciously and don't overload the system.

> Area 2

Same as with Kueue: resource tracking must consider it. The Cluster Autoscaler gets that for free by using the in-tree code.

Reply:
Accessing accelerators is always resource constrained. Given that our stance is that we should not use resource tracking for ResourceQuota/Kueue/Autoscaling, I am concerned that someone could use this path to get around typical resource quota requests.

I see that the scope of this is to access existing devices. Is there anything in place to make sure that the device is used? I worry that this could be used to potentially abuse access to devices if scheduler/kueue/resourcequota are bypassed but there are no restrictions in place to make sure that there is a request to this device already.

Reply:
> Accessing accelerators is always resource constrained. Given that our stance is that we should not use resource tracking for ResourceQuota/Kueue/Autoscaling, I am concerned that someone could use this path to get around typical resource quota requests.

But they would need privileged access to the cluster to do that. Are we really worried about people with admin rights abusing the cluster?

> Is there anything in place to make sure that the device is used? 

No. Why should it have to be in use? There's nothing wrong with collecting information about idle devices.

Reply:
I don't think my concerns are for alpha so it shouldn't block this.

I am mostly worried for autoscaling users being able to access new GPUS and we are essentially dropping kube protections. But we would have cloud/physical quota limits to guard against something accessing GPUs nodes if they are not actually allowed.